org.eclipse.xtext.xbase.testing/src/org/eclipse/xtext/xbase/testing/OnTheFlyJavaCompiler2.java

```diff
@@ -0,0 +1,202 @@
+				message.newLine();
+				message.append("\t");
+				message.append(Joiner.on("\n").join(result.getCompilationProblems()), "\t");
+				message.newLineIfNotEmpty();
+				throw new IllegalStateException(message.toString(), e);
+			} else {
+				throw Exceptions.sneakyThrow(t);
+			}
+		}
+	}
+
+	private String toJavaFile(String string) {
+		return string.replace('.', '/') + ".java";
+	}
+
+	public Map<String, Class<?>> compileToClasses(Map<String, String> sources) {
+		InMemoryJavaCompiler.Result result = inMemoryCompiler.compile(((JavaSource[]) Conversions.unwrapArray(
+				Iterables.transform(sources.entrySet(), e -> new JavaSource(toJavaFile(e.getKey()), e.getValue())),
+				JavaSource.class)));
+		try {
+			if (Iterables.any(result.getCompilationProblems(), CategorizedProblem::isError)) {
+				StringConcatenation message = new StringConcatenation();
+				message.append("Java code compiled with errors:");
+				message.newLine();
+				message.append(Joiner.on("\n")
+						.join(Iterables.filter(result.getCompilationProblems(), CategorizedProblem::isError)));
+				message.newLineIfNotEmpty();
+				message.newLine();
+				message.append("Code was:");
+				message.newLine();
+				message.append("=========");
+				message.newLine();
+				message.append(Joiner.on("\n=========\n").join(sources.values()));
+				message.newLineIfNotEmpty();
+				message.append("=========");
+				message.newLine();
+				throw new IllegalArgumentException(message.toString());
+			}
+			ClassLoader classLoader = result.getClassLoader();
+			return IterableExtensions.toMap(
+					Iterables.transform(sources.keySet(), (String it) -> {
+						try {
+							return classLoader.loadClass(it);
+						} catch (ClassNotFoundException e) {
+							throw Exceptions.sneakyThrow(e);
+						}
+					}), Class::getName);
+		} catch (Throwable t) {
+			if (t instanceof ClassNotFoundException) {
+				ClassNotFoundException e = (ClassNotFoundException) t;
+				StringConcatenation message = new StringConcatenation();
+				message.append(e.getMessage());
+				message.append(" ");
+				message.newLineIfNotEmpty();
+				message.append("source :");
+				message.newLine();
+				message.append("\t");
+				message.append(sources, "\t");
+				message.newLineIfNotEmpty();
+				message.newLine();
+				message.append("PROBLEMS : ");
+				message.newLine();
+				message.append("\t");
+				message.append(IterableExtensions.join(result.getCompilationProblems(), "\n"), "\t");
+				message.newLineIfNotEmpty();
+				throw new IllegalStateException(message.toString(), e);
+			} else {
+				throw Exceptions.sneakyThrow(t);
+			}
+		}
+	}
+
+	public <RT extends Object> Function0<RT> createFunction(String expression, Class<RT> returnType) {
+		try {
+			StringConcatenation sourceCode = new StringConcatenation();
+			sourceCode.append("public class __Generated implements org.eclipse.xtext.xbase.lib.Functions.Function0<");
+			sourceCode.append(returnType.getName());
+			sourceCode.append("> {");
```

Comment:
This may read better with
```
sourceCode.append(String.format("public class __Generated implements org.eclipse.xtext.xbase.lib.Functions.Function0<%s> {", returnType.getName())
```
If you agree there are other line candidates.